packages/drivers/odsp-driver/src/odspDocumentService.ts

```diff
@@ -303,26 +286,126 @@ export class OdspDocumentService implements IDocumentService {
+                            this._policies = {...this._policies,storageOnly: true};
+                            throw new DeltaStreamConnectionForbiddenError(code, { driverVersion });
+                        default:
+                            continue;
+                    }
+                }
+            }
+            throw e;
+        });
+    }
+
+    private async joinSessionCore(
+        requestSocketToken: boolean,
+        options: TokenFetchOptionsEx,
     ): Promise<ISocketStorageDiscovery> {
-        const executeFetch = async () =>
-            fetchJoinSession(
-                this.odspResolvedUrl,
-                "opStream/joinSession",
-                "POST",
-                this.mc.logger,
-                this.getStorageToken,
-                this.epochTracker,
-                requestSocketToken,
-                options,
-                this.hostPolicy.sessionOptions?.unauthenticatedUserDisplayName,
-            );
+        const disableJoinSessionRefresh = this.mc.config.getBoolean("Fluid.Driver.Odsp.disableJoinSessionRefresh");
+        const executeFetch = async () => {
+            return {
+                    entryTime: Date.now(),
+                    joinSessionResponse: await fetchJoinSession(
+                        this.odspResolvedUrl,
+                        "opStream/joinSession",
+                        "POST",
+                        this.mc.logger,
+                        this.getStorageToken,
+                        this.epochTracker,
+                        requestSocketToken,
+                        options,
+                        disableJoinSessionRefresh,
+                        this.hostPolicy.sessionOptions?.unauthenticatedUserDisplayName,
+                    )
+            };
+        };
+
+        let response = await this.cache.sessionJoinCache.addOrGet(this.joinSessionKey, executeFetch);
+        // If the response does not contain refreshSessionDurationSeconds, then treat it as old flow and let the
+        // cache entry to be treated as expired after 1 hour.
+        response.joinSessionResponse.refreshSessionDurationSeconds =
+            response.joinSessionResponse.refreshSessionDurationSeconds ?? 3600;
+        let refreshSessionDurationSeconds: number | undefined =
+            response.joinSessionResponse.refreshSessionDurationSeconds;
+        let refreshAfterDeltaMs =
+            this.calculateJoinSessionRefreshDelta(response.entryTime, refreshSessionDurationSeconds);
+        // This means that the cached entry has expired(This should not be possible if the response is fetched
+        // from the network call). In this case we remove the cached entry and fetch the new response.
+        if (refreshAfterDeltaMs <= 0) {
+            this.cache.sessionJoinCache.remove(this.joinSessionKey);
+            response = await this.cache.sessionJoinCache.addOrGet(this.joinSessionKey, executeFetch);
```

Comment:
This logic is repeated (see lines 361+).
If you move this logic into executeFetch callback, then you would not need to repeat it.
Basically you can capture start time, do await fetchJoinSession, and then recalculate end time and return only end time & response from executeFetch, and thus logic would be in one place.

Reply:
Simplified logic futher more.